src/Storages/StorageS3.cpp

```diff
@@ -1757,7 +2115,54 @@ namespace
                 }
 
                 int zstd_window_log_max = static_cast<int>(getContext()->getSettingsRef().zstd_window_log_max);
-                auto impl = std::make_unique<ReadBufferFromS3>(configuration.client, configuration.url.bucket, current_key_with_info->key, configuration.url.version_id, configuration.request_settings, getContext()->getReadSettings());
+                std::unique_ptr<ReadBufferFromFileBase> impl;
+                if (!current_key_with_info->info)
+                    current_key_with_info->info = S3::getObjectInfo(
+                        *configuration.client,
+                        configuration.url.bucket,
+                        current_key_with_info->key,
+                        configuration.url.version_id,
+                        configuration.request_settings);
+                LOG_DEBUG(
+                    &Poco::Logger::get("Schema Inference 2"),
+                    "File info: key: {} size: {} archive path: {} format: {}",
+                    current_key_with_info->key,
+                    current_key_with_info->info->size,
+                    current_key_with_info->path_in_archive.has_value() ? current_key_with_info->path_in_archive.value() : "Empty path",
+                    format.has_value() ? format.value() : "No format");
+                if (!current_key_with_info->path_in_archive.has_value())
+                {
+                    impl = std::make_unique<ReadBufferFromS3>(
+                        configuration.client,
+                        configuration.url.bucket,
+                        current_key_with_info->key,
+                        configuration.url.version_id,
+                        configuration.request_settings,
+                        getContext()->getReadSettings());
+                }
+                else
+                {
+                    assert(current_key_with_info->archive_holder->getArchiveReader());
+                    // std::unique_ptr<ReadBufferFromFileBase> inter_buf = std::make_unique<ReadBufferFromS3>(
+                    //     configuration.client,
+                    //     configuration.url.bucket,
+                    //     current_key_with_info->key,
+                    //     configuration.url.version_id,
+                    //     configuration.request_settings,
+                    //     getContext()->getReadSettings());
+                    // // LOG_DEBUG(&Poco::Logger::get("Debug 100"), "");
+                    // // ;
+                    // LOG_DEBUG(&Poco::Logger::get("Debug 101"), "");
+                    impl = current_key_with_info->archive_holder->getArchiveReader()->readFile(
```

Comment:
Same problem here, we will wrap it with compression method again:
![image](https://github.com/ClickHouse/ClickHouse/assets/3684697/5c8e3b6d-91c1-49a0-a556-04a8aaaf4296)